pynq/lib/dma.py

```diff
@@ -110,20 +119,30 @@ def transfer(self, array):
         ----------
         array : ContiguousArray
             An xlnk allocated array to be transferred
+        start : integer (optional, default 0)
+             Offset into array to start
+        nbytes : integer (optional, default 0)
+             Number of bytes to transfer.  0 means everything after the starting point.
 
         """
-        if array.nbytes > self._size:
-            raise ValueError('Transferred array is {} bytes, which exceeds '
-                             'the maximum DMA buffer size {}.'.format(
-                              array.nbytes, self._size))
         if not self.running:
             raise RuntimeError('DMA channel not started')
         if not self.idle and not self._first_transfer:
             raise RuntimeError('DMA channel not idle')
+        if nbytes == 0:
+            nbytes = array.nbytes - start
+        if (nbytes > self._size):
+            raise ValueError('Transfer size is {} bytes, which exceeds '
+                             'the maximum DMA buffer size {}.'.format(
+                              nbytes, self._size))
+        if self._dre and ((array.pyhsical_address + start) % 0x40) != 0:
```

Comment:
Checking the data-sheet indicates that the alignment for non-SG ports is dependent on the AXI Master width rather than being fixed to 64 bytes.

Reply:
Hmm, you're right.. "source address must be [MM2S|S2MM] Memory Map data width aligned".  So I guess that means the "c_m_axi_[s2mm|mm2s]_data_width" property?

(And it looks like I got the polarity of the 'self._dre' test backwards.  oops..)